File: eel-core/src/test/kotlin/io/eels/RowUtilsTest.kt
package io.eels

import io.eels.schema.Field
import io.eels.schema.Schema
import io.kotlintest.specs.StringSpec

class RowUtilsTest : StringSpec() {
  init {
    "rowAlign should reorder in line with target schema" {
      val row = Row(Schema(Field("a"), Field("b"), Field("c")), listOf("aaaa", "bbb", "ccc"))
      val targetSchema = Schema(Field("c"), Field("b"))
      RowUtils.rowAlign(row, targetSchema) shouldBe Row(Schema(Field("c"), Field("b")), listOf("ccc", "bbb"))
    }
    "rowAlign should lookup missing data" {
      val row = Row(Schema(Field("a"), Field("b"), Field("c")), listOf("aaaa", "bbb", "ccc"))
      val targetSchema = Schema(Field("c"), Field("d"))
      RowUtils.rowAlign(row, targetSchema, mapOf("d" to "ddd")) shouldBe Row(Schema(Field("c"), Field("d")), listOf("ccc", "ddd"))
    }
    "rowAlign should throw an error if a field is missing" {
      val row = Row(Schema(Field("a"), Field("b"), Field("c")), listOf("aaaa", "bbb", "ccc"))
      val targetSchema = Schema(Field("c"), Field("d"))
      shouldThrow<IllegalStateException> {
        RowUtils.rowAlign(row, targetSchema)
      }
    }
  }
}

File: eel-core/src/test/scala/io/eels/RowUtilsTest.scala
package io.eels

import io.eels.schema.{BooleanType, DoubleType, Field, StringType, StructType}
import org.scalatest.{Matchers, WordSpec}

class RowUtilsTest extends WordSpec with Matchers {
  "RowUtils.rowAlign" should {
    "rowAlign should reorder in line with target schema" in {
      val row = Row(StructType(Field("a"), Field("b"), Field("c")), "aaaa", "bbb", "ccc")
      val targetSchema = StructType(Field("c"), Field("b"))
      RowUtils.rowAlign(row, targetSchema) shouldBe Row(StructType(Field("c"), Field("b")), "ccc", "bbb")
    }
    "rowAlign should lookup missing data" in {
      val row = Row(StructType(Field("a"), Field("b"), Field("c")), "aaaa", "bbb", "ccc")
      val targetSchema = StructType(Field("c"), Field("d"))
      RowUtils.rowAlign(row, targetSchema, Map("d" -> "ddd")) shouldBe Row(StructType(Field("c"), Field("d")), "ccc", "ddd")
    }
    "rowAlign should throw an error if a field is missing" in {
      val row = Row(StructType(Field("a"), Field("b"), Field("c")), "aaaa", "bbb", "ccc")
      val targetSchema = StructType(Field("c"), Field("d"))
      intercept[RuntimeException] {
        RowUtils.rowAlign(row, targetSchema)
      }
    }
  }

  "RowUtils.coerce" should {
    "coerce values to match types" in {
      val schema = StructType(Field("a", StringType), Field("b", DoubleType), Field("c", BooleanType))
      val row = Row(schema, Vector(1, "1.4", "true"))
      RowUtils.coerce(row) shouldBe Row(schema, "1", 1.4D, true)
    }
  }
}